bayes_opt1/util.py

```diff
@@ -0,0 +1,547 @@
+    p = optimizer._space.bounds.shape[0] # size of state space (e.g. if we are working on R^3, n = 3)
+    init = [np.zeros(n_grid)] * p # list of p vectors, each of lenght n_grid   
+    x_obs_temp = np.array([[res["params"]] for res in optimizer.res]) 
+    y_obs = np.array([res["target"] for res in optimizer.res])
+    # x_obs needs to be np.array of size (n,p)
+    # y_obs needs to be np.array of size (n,)
+    # see https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html
+    x_obs = x_obs_to_array(x_obs_temp, p)
+
+    for i in range(p): 
+        init[i] = np.linspace(optimizer._space.bounds[i,0], optimizer._space.bounds[i,1], n_grid) # works \forall p
+
+    grid = np.meshgrid(*init) # list with p matrices n_grid x n_grid
+    for g in range(len(grid)):
+        grid[g] = grid[g].reshape(-1,1)
+    grid = np.stack(grid, axis = -1)
+    grid=grid[:,0,:] # array of shape (n_grid*n_grid, p): each row is composed of one element from init[0] and one element from init[1]
+    
+    temp = find_mean_max(x_obs,y_obs,gp,grid)
+    mean_nstar = temp[0]
+
+    grad_temp = []
+    mean, std= gp.predict(x.reshape(1,-1), return_std=True)
+    for j in range(J):
+        print("j = ",j+1)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            temp = find_mean_max(x_obs,y_obs,gp,grid,mean,std,x)
+            mean_n1star = temp[0]
+            x_star_hat = grid[temp[1]].reshape(1,-1)
+            def func(w):
+                return mu_x_star_hat(x_obs,y_obs,gp,x_star_hat,mean,std,w)
+            
+            grad_temp.append(nd.Gradient(func)(x))
+    grad = np.zeros(p)
+    for i in grad_temp:
+        grad = grad + i
+    grad = grad/J
+    return grad
+
+
+
+# Implementation algorithm 3 (Frazier): Knowledge gradient 
+def minimize_kg_sgd(R, T, a, p, pbounds, optimizer, gp):
+
+    alpha = np.zeros(T)
+    KG = np.zeros(R)
+    temp = our_lhs(p,R,pbounds)
+    
+    if p == 1:
+        
+        x0 = np.zeros((T,R)) 
+        
+        for r in range(R):
+            # choose x0(r) uniformly random from A
+            t=0
+
+            x0[t,r] = temp[r]
+
+            for t in range(T-1):
+                #print(x0[(t+1)*r].reshape(1,-1))
+                G = _kg4(x0[t,r], optimizer, gp, n_grid = 50, J = 5)
+                alpha[t+1] = a/(a+t+1)
+                x0[t+1,r] = x0[t,r] + alpha[t+1]*G
+            #print(type(x0[T-1,r]))        
+            KG[r] = _kg2(x0[T-1,r], optimizer, gp, n_grid = 50, J = 5)
+
+
+    else:
+        x0 = []  
+
+        for r in range(R):
+            # choose x0(r) uniformly random from A
+            t=0
+
+            x0.append(temp[r].reshape(-1,1)) 
+
+            for t in range(T-1):
+                print(x0[(t+1)*r].reshape(1,-1))
+                G = _kg4(x0[(t+1)*r].reshape(1,-1), optimizer, gp, n_grid = 50, J = 5)
+                alpha[t+1] = a/(a+t+1)
+                x0.append(x0[(t+1)*r] + alpha[t+1]*G)
```

Comment:
Non siamo molto sicuri che l'indice di x0 sia corretto... potrebbe controllarlo?

Reply:
Effettivamente gli indici di `x0` sono sbagliati, ma rileggendo queste righe ho notato altri errori. Anzitutto `range(x)` contiene `x` numeri che vanno da 0 fino a `x-1`, non `x`, quindi voi ora state facendo un'iterazione in meno. Inoltre anche quello che chiamate `alpha[t+1]` dovrebbe in realtà essere `alpha[t]`, perché in Python gli indici partono da zero! (Immagino che questi due errori siano correlati.)

In conclusione: potreste inizializzare `x0 = np.zeros((T+1,R)) ` invece che `T` e mettere alla riga 0 i valori iniziali. Poi proseguite con il `range()` loop ma partendo da 1 invece che da 0 (vedete la documentazione). Così facendo potete tenere gli stessi indici del paper senza confondervi!

Reply:
Abbiamo modificato gli indici.. ora sono giusti? In particolare abbiamo il dubbio in x0.append(...)